### PR TITLE
fix retention size policy delete too much ledgers

### LINKED ISSUE #11241: Retention size policy may delete too much ledgers
**Describe the bug**
If namespaces is enable Retention size policy, broker may delete too much ledgers if all cursors have no backlog.

Assume that we have a topic with retentino size is 5MB, and retention time is -1.
And this topic have 10 ledgers  in `ManagedLedger` and each ledger has 1MB size,  and all cursors has no backlog.
when the retention size policy works, i't will delete the first 9 ledgers.

**To Reproduce**
1. Create a topic , set retention size policy to a proper size and set retention time to -1(disable )
2. Create a subscritption(cursor) for this topic
3. send some messages to make sure ledgers rollover several times
4. consume all the messages from the topic using the created subscription
5. broker will delete all the ledgers besides the last ledger

**Expected behavior**
Broker should maintain proper size of message according to the retention size instead of deleting all acked ledgers

**Screenshots**

**Desktop (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by aloyszhang**
Fixes #11241
### Motivation
This pull request fix the error that retention size policy delete too much ledgers when all cursors has no backlog

### Modifications

1.  generate `ledgersToDelete` by checking ledger size one by one instead of `TOTAL_SIZE_UPDATER`
2. refactor the logic of generate `ledgersToDelete`
3. add test code

### Verifying this change
This change added tests and can be verified as follows:
- add test in `ManagedLedgerTest.testRetentionSize()`

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Review by merlimat — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 1]
```diff
@@ -2369,33 +2375,52 @@ void internalTrimLedgers(boolean isTruncate, CompletableFuture<?> promise) {
...
                 if (ls.getLedgerId() == currentLedger.getId()) {
                     log.debug("[{}] Ledger {} skipped for deletion as it is currently being written to", name,
                             ls.getLedgerId());
                     break;
-                } else if (expired || isTruncate) {
-                    log.debug("[{}] Ledger {} has expired or be truncated, expired is {}, isTruncate is {}, ts {}", name, ls.getLedgerId(), expired,  isTruncate, ls.getTimestamp());
+                }
+                // if truncate, all ledgers besides currentLedger are going to be deleted
+                if (isTruncate){
+                    log.debug("[{}] Ledger {} has been truncated, , ts {}", name, ls.getLedgerId(), ls.getTimestamp());
```
Comment (codelipenghui) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2392]:
```suggestion
                    log.debug("[{}] Ledger {} has been truncated with ts {}", name, ls.getLedgerId(), ls.getTimestamp());
```

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 2]
```diff
@@ -2369,33 +2375,52 @@ void internalTrimLedgers(boolean isTruncate, CompletableFuture<?> promise) {
...
+                            retentionSizeQuotaMet = true;
+                            ledgersToDelete.add(ls);
+                            continue;
+                        }
+                    } else {
+                        retentionSizeQuotaMet = true;
+                    }
+                }
+
+                if (hasLedgerRetentionExpired(ls.getTimestamp())) {
```
Comment (codelipenghui) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2417]: I'm not sure why the PR can fix the problem. Looks the problem is the existing logic is if the ledger over retention quota or expired, the ledger will be added into the `ledgersToDelete`, the correct logic is if the ledger over retention quota and expired, we should add it into the `ledgersToDelete` right?

But looks like the new change still adds the ledger to the `ledgersToDelete` after over retention quota without checking the expired.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java [hunk 3]
```diff
@@ -1970,6 +1971,46 @@ public void testInfiniteRetention() throws Exception {
...
+        final int messageSize = 1048576;
+        char[] data = new char[messageSize];
+        Arrays.fill(data, 'a');
+        byte [] message = new String(data).getBytes(Encoding);
+
+        @Cleanup("shutdown")
+        ManagedLedgerFactory factory = new ManagedLedgerFactoryImpl(metadataStore, bkc);
+        ManagedLedgerConfig config = new ManagedLedgerConfig();
+        config.setRetentionSizeInMB(retentionSizeInMB);
+        config.setMaxEntriesPerLedger(1);
```
Comment (codelipenghui) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java:1989]: We have a minRolloverTime limitation by default 10min,  Set max 1 entry per ledger will not trigger the ledger rollover.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java [hunk 4]
```diff
@@ -1970,6 +1971,46 @@ public void testInfiniteRetention() throws Exception {
...
+        config.setRetentionTime(1, TimeUnit.HOURS);
+
+
+        ManagedLedgerImpl ml = (ManagedLedgerImpl) factory.open("retention_size_ledger", config);
+        ManagedCursor c1 = ml.openCursor("c1");
+        Position position = null;
+        for (int i = 0; i < totalMessage; i++) {
+            position = ml.addEntry(message);
+        }
+        List<Entry> entryList = c1.readEntries(totalMessage);
```
Comment (codelipenghui) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java:1999]: We should check how many ledgers for the ManagedLedger for now, it should be 10 ledgers.

**Review by codelipenghui — CHANGES_REQUESTED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 2 above)
Comment (aloyszhang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2417]: >  the correct logic is if the ledger over retention quota and expired, we should add it into the ledgersToDelete right?

I think the logic of  retention policy means  either over retention quota or expired, the ledger should be delete.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 1 above)
Comment (aloyszhang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2392]: sure, thanks

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java (hunk 3 above)
Comment (aloyszhang) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java:1989]: The default `minimumRolloverTimeMs` is 0 

 https://github.com/apache/pulsar/blob/master/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerConfig.java?#L53

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java (hunk 4 above)
Comment (aloyszhang) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java:1999]: sure

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 2 above)
Comment (codelipenghui) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2423]: Oh. sorry, My mistake.

**Review by codelipenghui — APPROVED** [review]

**Comment by codelipenghui** [comment]
Nice catch @aloyszhang

**Pull request merged**